performance_manager/lib/rt_vehicle_positions.py

```diff
@@ -0,0 +1,277 @@
+    new events will be merged with existing events in a 5 minutes window
+    surrounding the year/month/day/hour value found in path of parquet files
+    """
+
+    start_window_ts = get_utc_timestamp(folder)
+    timestamp_to_pull_min = start_window_ts - 300
+    timestamp_to_pull_max = start_window_ts + 300 + 3600
+    get_db_events = select(
+        (
+            VehiclePositionEvents.id,
+            VehiclePositionEvents.hash,
+            VehiclePositionEvents.timestamp_start,
+            VehiclePositionEvents.timestamp_end,
+            VehiclePositionEvents.vehicle_id,
+        )
+    ).where(
+        (VehiclePositionEvents.timestamp_start > timestamp_to_pull_min)
+        & (VehiclePositionEvents.timestamp_end < timestamp_to_pull_max)
+    )
+    with session.begin() as s:  # type: ignore
+        if s.execute(get_db_events).first() is None:
+            logging.info(f"No merge events found in database.")
+            merge_events = new_events.sort_values(
+                by=["vehicle_id", "timestamp_start"]
+            )
+        else:
+            # Join records from db with records from parquet
+            # db records should contain 'index' column that does not exist in parquet dataset
+            merge_events = pandas.concat(
+                [
+                    pandas.DataFrame(
+                        [r._asdict() for r in s.execute(get_db_events)]
+                    ),
+                    new_events,
+                ]
+            ).sort_values(by=["vehicle_id", "timestamp_start"])
+
+    # Identify records that are continuing from existing db
+    # If such records are found, update timestamp_end with latest value
+    first_of_consecutive_events = (
+        merge_events["hash"] - merge_events["hash"].shift(-1) == 0
+    )
+    last_of_consecutive_events = (
+        merge_events["hash"] - merge_events["hash"].shift(1) == 0
+    )
+    merge_events["timestamp_end"] = numpy.where(
+        first_of_consecutive_events,
+        merge_events["timestamp_end"].shift(-1),
+        merge_events["timestamp_end"],
+    )
+
+    existing_was_updated_mask = (
+        ~(merge_events["id"].isna()) & first_of_consecutive_events
+    )
+
+    existing_to_del_mask = (
+        ~(merge_events["id"].isna()) & last_of_consecutive_events
+    )
+
+    logging.info(
+        f"Size of existing update df: {existing_was_updated_mask.sum():,}"
+    )
+    logging.info(f"Size of existing delete df: {existing_to_del_mask.sum():,}")
+
+    """
+    new events that will be inserted into db table
+    """
+    new_to_insert_mask = (
+        merge_events["id"].isna()
+    ) & ~last_of_consecutive_events
+    logging.info(f"Size of new insert df: {new_to_insert_mask.sum():,}")
+
+    """
+    new events that were used to update existing records, will not be used
+    """
+    new_to_drop_mask = (merge_events["id"].isna()) & last_of_consecutive_events
+    logging.info(f"Size of new being dropped df: {new_to_drop_mask.sum():,}")
```

Comment:
All of these masks identify which records will be UPDATED/INSERTED/DELETED in the database.